### ISSUE TITLE
Insidious random data/memory corruption bug causing incorrect computation and training divergence

### ISSUE THREAD

**Issue by xuancong84** [Theano/Theano#4770]
It seems that sometimes by chance, Theano's (for all versions including bleeding-edge) internal memory state can get corrupted silently, with all subsequent training/testing operations produces erroneous results without throwing any exceptions/warnings. The error will accumulate until some point when the training starts to always diverge. The problem can be solved by aborting the current process, reloading the last-known good model and resuming training.

My code does iterative training, it saves the model at every iteration. After every iteration, it checks the loss function, if it increases, then load the previous iteration's model and reduce learning rate by half. The bug starts to emerge randomly somewhere in between 20-80th iteration (for my code), starting from which the loss keeps increasing after every iteration despite the fact that the learning rate is halved each time and the model keeps falling back to the previous iteration. This process continues until the learning rate becomes so small, <1e-200, and thus, there is no change in the weight or the loss function.

For example,

```
> 1310.31767385 Training acc= 0.971412992678      
> Testing Iter 50 :                               
> Test accuracy = 0.40635451505                   
> Training Iter 51 :                              
> 1213.0716554 Training acc= 0.976086617853       
> Testing Iter 51 :                               
> Test accuracy = 0.407190635452                  
> Training Iter 52 :                              
> 1248.2027879 Training acc= 0.969699330114       
> Testing Iter 52 :                               
> Test accuracy = 0.409698996656                  
> Falling back to Iter 51 ; learn_rate= 0.00125   
> Training Iter 52 :                              
> 1394.18119661 Training acc= 0.964246767409      
> Testing Iter 52 :                               
> Test accuracy = 0.418060200669                  
> Falling back to Iter 51 ; learn_rate= 0.000625  
> Training Iter 52 :                              
> 2215.10956366 Training acc= 0.937139741393      
> Testing Iter 52 :                               
> Test accuracy = 0.413879598662                  
> Falling back to Iter 51 ; learn_rate= 0.0003125 
> Training Iter 52 :                              
> 3660.30786752 Training acc= 0.897647608662      
> Testing Iter 52 :                               
> Test accuracy = 0.413879598662                  
> Falling back to Iter 51 ; learn_rate= 0.00015625
> Training Iter 52 :                              
> 5290.97243645 Training acc= 0.863919613647      
> Testing Iter 52 :                               
> Test accuracy = 0.420568561873                  
> Falling back to Iter 51 ; learn_rate= 7.8125e-05
```

I have explicitly checked that there is no corruption in the data, i.e., no NAN, INF, underflow/overflow, in all saved model files, regardless of whether the loss increases or decreases. Moreover, there is no bug in saving and loading models, because in previous iterations (before 51), the falling back mechanism has been triggered several times and it was working correctly.

The files are at http://translate.i2r.a-star.edu.sg/theano-bug.zip
The command line to produce the bug is in do.sh
Moreover, I have also translated exactly the same expression graph into Tensorflow, and there is no such divergence issue, suggesting that it is unlikely to be a bug in my program, see the 4th post for more details.

**Comment by nouiz** [comment]
Can you try on another computer?

We got some strange behavior like this due to hardware problem, like bad
memory or GPU with problems.

Was this on CPU or GPU?

**Comment by xuancong84** [comment]
I have tried on 4 unrelated computers, two from my company (my Desktop and our server), two from my university. All have the same behaviour. It does not seem to be a hardware problem.

I have tried on cpu only. Because GPU is 3 times slower due to the scan function.

I am pretty sure it is a theano's problem from all my experiments and observations.

**Comment by xuancong84** [comment]
I have translated the entire expression graph into Tensorflow and have tested. The results show that although Theano runs much faster (5-6 times) than Tensorflow (both tested on CPU as GPU is 3-4 times slower due to the "scan" function), Theano's computation is ultimately incorrect.

In the beginning, given the same NN expression graph, same random initialization, same initial learning rate, same gradient descent, both Theano and Tensorflow have about the same initial loss, training acc and testing acc, this is expected:
```
**Tensorflow:**
Training Iter 0 :
[ 15195.79845379] Training acc= 0.514098769279
Testing Iter 0 :
Test accuracy = 0.473244147157

**theano:**
Training Iter 0 :
15004.9402727 Training acc= 0.517759775666
Testing Iter 0 :
Test accuracy = 0.477424749164
```

However, very soon, they start to differ:
```
**Tensorflow:**
Training Iter 17 :
[ 7203.33795861] Training acc= 0.782598535597
Testing Iter 17 :
Test accuracy = 0.405518394649
Training Iter 18 :
[ 6451.10572737] Training acc= 0.80721296152
Testing Iter 18 :
Test accuracy = 0.422240802676
Training Iter 19 :
[ 5732.66933589] Training acc= 0.830581087397
Testing Iter 19 :
Test accuracy = 0.404682274247

**Theano:**
Training Iter 17 :
12141.549198 Training acc= 0.590746222153
Testing Iter 17 :
Test accuracy = 0.454013377926
Training Iter 18 :
11884.4289652 Training acc= 0.601962922574
Testing Iter 18 :
Test accuracy = 0.446488294314
Training Iter 19 :
11606.5018466 Training acc= 0.612634366724
Testing Iter 19 :
Test accuracy = 0.444816053512
```

Interestingly, if you look at the loss vs acc ratio, they are more or less consistent
```
**Tensorflow:**
Training Iter 15 :
[ 8676.95984604] Training acc= 0.732201277458
Testing Iter 15 :
Test accuracy = 0.411371237458
Training Iter 16 :
[ 7961.80853327] Training acc= 0.753154696993
Testing Iter 16 :
Test accuracy = 0.401337792642
Training Iter 17 :
[ 7203.33795861] Training acc= 0.782598535597
Testing Iter 17 :
Test accuracy = 0.405518394649

**Theano:**
Training Iter 29 :
8233.35114998 Training acc= 0.730253933635
Testing Iter 29 :
Test accuracy = 0.403010033445
Training Iter 30 :
7861.4908714 Training acc= 0.742872721608
Testing Iter 30 :
Test accuracy = 0.405518394649
Training Iter 31 :
7424.77473705 Training acc= 0.75627044711
Testing Iter 31 :
Test accuracy = 0.364548494983
```

The above shows that for both Theano and TF, at about the same loss, their respective training and test acc are about the same. The reason is because this is determined by the NN model property, which should be independent of the toolkit.

This suggests that my initial conjecture is correct, Theano's random internal memory/data corruption behaves like cumulative-poisoning: in the beginning, when the corruption is not severe enough, the training can still converge, however, it converges slower than it should; at some point, when the degree of corruption becomes too serious, the training starts to diverge, no matter how small the learning rate is. Unfortunately, the corruption does not lie in the external accessible data, all saved/loaded data arrays are intact (as I checked, there is no NAN, INF, overflow/underflow in all variables, reloading the model and resume training works correctly). Rather, the corruption lies in the internal data or even bug in the compiled C++ code, which is inaccessible from Python. Therefore, whenever this happens, we can do nothing but have to abort the program, reload the last-known good model and resume the training.

**Comment by lamblin** [comment]
Does this happen as well with `optimizer=fast_compile`?

**Comment by xuancong84** [comment]
@lamblin Yes, **not only does it happen to both optimizer=fast_compile and optimizer=fast_run, but it also happens to all DEVICE=gpu/cpu/cuda, and it also happens to both float32 and float64**.

The good thing is that for optimizer=fast_compile and optimizer=fast_run, (floatX=double) within the first 20 iterations, all the training loss, training acc and test acc are exactly the same up to all displayed 12 significant figures. This shows a very good consistency between the C function and Python function.

However, to me, Theano is still considered as a piece of universal garbage. It computes super fast, but its calculations are all wrong, then what is the point of being so fast? -:(

**Comment by ozancaglayan** [comment]
If I'm not mistaken, you did not even set the random seed in your code, how can you compare the first 20 iterations of different runs?

**Comment by nouiz** [comment]
I would recommend to check @ozancaglayan comment first. He also found something else in a different issue. Can you try with this Theano flag and check if you get different result? 

optimizer_including=alloc_empty_to_zeros

**Comment by xuancong84** [comment]
@ozancaglayan  Sorry, I did not set random seed in the initial version. So I am not making an exact comparison here. But despite that, the bug is still valid. Even if the initial weights are different by random, the general trend for convergence should be about the same and not very different.

@nouiz Unfortunately, after I add "optimizer_including=alloc_empty_to_zeros" into Theano flag, the bug is still there, it still diverges after some iteration. And the speed of convergence is still much slower than tensorflow -:(

In this program, the bug might be due to the fact that I have used the Theano scan function extensively. We all hope that without using scan, Theano should do a good job. Unfortunately, that is not true. I just found another bug in which the scan function is not used, but cpu, gpu and cuda all give different results. Will be posted into another issue.

**Comment by ozancaglayan** [comment]
Can you create a github repository with your test code?

**Comment by as1986** [comment]
@xuancong84 do you have an update on this bug? Because I seem to also have run into the same problem...

**Comment by nouiz** [comment]
Be sure to update Theano to the dev version: 

http://www.deeplearning.net/software/theano/install.html#bleeding-edge-install-instructions

It fixed @ozancaglayan problems.

**Comment by xuancong84** [comment]
@nouiz What is @ozancaglayan 's problem? I have tested, on my test code it still diverges after some random epoches :(
Btw, do you guys use Keras? Is it possible that for some code, Keras produces different results when running on CUDA, CPU and GPU, but Theano produces the same results on CUDA, CPU and GPU? I am very confused.

**Comment by nouiz** [comment]
What is the status of this? Is it fixed with the master of Theano? There have been fix related.

If we can't reproduce it, I'll close this.

Just to be sure, CPU will always return small different result compared to a GPU, as the order of addition/multiplication is different and this change the results.

**Comment by xuancong84** [comment]
@nouiz It is a tough bug. I don't think it is fixed. You may close this thread since I no longer use Theano any more. Too disappointed -:(

**Comment by lamblin** [comment]
I tried to reproduce the issue of non-determinism, and I was not able to. Here is what I did:
- I set a seed so the initialization is consistent across runs (`numpy.random.seed(77)`)
- I used `floatX=float64,device=cpu`
- Running the job multiple times, with and without `optimizer_including=alloc_empty_to_zeros` always "diverged" (training accuracy going down) at the same point (iteration 48 falls back repeatedly to iteration 47)
- Restarting from an earlier checkpoint with the same *current* learning rate (for instance, iteration 46 with learning rate 0.005 rather than the initial learning rate of 0.01) gives again exactly the same numbers, and diverges in the same way.
- However, restarting with the original learning rate produces different results, as expected, that can escape that local minimum in training error.

Therefore, I don't think there is a memory corruption bug, or anything unexpectedly non-deterministic going on with scan, at least at the moment.
